MMA8452.cpp

```diff
@@ -517,3 +517,25 @@ int8_t MMA8452::convertTo2sComplement(int8_t value)
 	value = (0xFF && ~value) + 1;
 	return value;
 }
+
+void MMA8452::getPitchRoll(float *pitch, float *roll)
+{
+	float Xg, Yg, Zg;
+
+	getAcceleration(&Xg, &Yg, &Zg);
+
+	//Low Pass Filter
+	fXg = Xg * alpha + (fXg * (1.0 - alpha));
+	fYg = Yg * alpha + (fYg * (1.0 - alpha));
+	fZg = Zg * alpha + (fZg * (1.0 - alpha));
+
+	*roll = (atan2(-fYg, fZg)*180.0) / M_PI;
+	*pitch = (atan2(fXg, sqrt(fYg*fYg + fZg*fZg))*180.0) / M_PI;
+}
+
+void MMA8452::initPitchRoll(float a)
```

Comment:
Is this needed at all? Could `fXg`, `fYG` and `fZg` not just need to be local variables in `getPitchRoll` & `alpha` be an argument to the function? If you forget to call `initPitchRoll`, `alpha` would not be set which would probably not work as expected. Also, it's not immediately clear what the purpose of `alpha` is; i suppose it's acting as range? It's not clear what value should be passed in (`0.5`?)

Reply:
Alpha indeed could be a parameter.
fXg, fYg and fZg need to be members or passed as reference (in/out) parameters to the function and be saved by the caller.
fXg, fYg and fZg are used as a low pass filter implemented as a weighted average of the current reading and the previous readings. It slows the response time but without it the result will be very jerky.  When Alpha == 1 the result is calculated without history (no low pass filter). When Alpha is close to zero (e.g. 0.1) the low pass filter is strong and the result is stable (albeit with a few hundred milliseconds delay). 0.5 is a good value for Alpha.

Reply:
Do you like this format better?
void MMA8452::getPitchRoll(float *pitch, float *roll, float alpha, float *fXg, float *fYg, float *fZg)
{
	float Xg, Yg, Zg;

	getAcceleration(&Xg, &Yg, &Zg);

	//Low Pass Filter
	*fXg = Xg * alpha + (*fXg * (1.0 - alpha));
	*fYg = Yg * alpha + (*fYg * (1.0 - alpha));
	*fZg = Zg * alpha + (*fZg * (1.0 - alpha));

	*roll = (atan2(- *fYg, *fZg)*180.0) / M_PI;
	*pitch = (atan2(*fXg, sqrt(*fYg * *fYg + *fZg * *fZg))*180.0) / M_PI;
}

Reply:
Oh, i see what you mean now! I've typically implemented lowpass as `v += (target - v) * a` so i missed this. This makes sense. Could you please add a comment on `alpha` though:

> alpha defines how quickly to react to changes. The value should be 0-1.

I would prefer to keep the signature smaller and not having to have the called keep track of these values. Was thinking though, since you're expected to call `init()` anyway, why not just set the moving average values (`fXg`/`fYg`/`fZg`) to `0` in there?